src/dmd/statementsem.d

```diff
@@ -1454,13 +1454,54 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
                 e = new VarExp(loc, r);
                 Expression einit = new DotIdExp(loc, e, idfront);
                 Statement makeargs, forbody;
+                bool ignoreRef = false; // If a range returns a non-ref front we ignore ref on foreach
+
+                Type tfront;
+                if (auto fd = sfront.isFuncDeclaration())
+                {
+                    if (!fd.functionSemantic())
+                        goto Lrangeerr;
+                    tfront = fd.type;
+                }
+                else if (auto td = sfront.isTemplateDeclaration())
+                {
+                    Expressions a;
+                    if (auto f = resolveFuncCall(loc, sc, td, null, tab, &a, 1))
+                        tfront = f.type;
+                }
+                else if (auto d = sfront.isDeclaration())
+                {
+                    tfront = d.type;
+                }
+                if (!tfront || tfront.ty == Terror)
+                    goto Lrangeerr;
+                if (tfront.toBasetype().ty == Tfunction)
+                {
+                    auto ftt = cast(TypeFunction)tfront.toBasetype();
+                    tfront = tfront.toBasetype().nextOf();
+                    if (!ftt.isref)
+                    {
+                        // .front() does not return a ref. We ignore ref on foreach arg.
+                        // see https://issues.dlang.org/show_bug.cgi?id=11934
+                        if (tfront.needsDestruction()) ignoreRef = true;
```

Comment:
I thought the point of the issue was to error (or `deprecate`) on `ref` if `front` doesn't return by `ref`?

Reply:
It seems that consensus is to simply allow silently it. Probably returning an error is going to break a lot of code and to complicate a lot of code with many corner cases (consider f.e. "each" template on Phobos).

Reply:
Hmm did you check how much code in Phobos would be affected?
As long as it only triggers a deprecation, we don't break _any_ code and this seems like one of those silent things that you really want your compiler to warn you about.

Reply:
Yes error will brake code but deprecation as said is going to complicate code with many corner cases. Give a look at "each" code. You probably have to duplicate the just-duplicated code. Moreover every single time you use foreach+ref with a range you have to check if front is a function that returns a ref and write a special case for it. It doesn't sound good to me. Anyway I can check deprecation against Phobos unit tests.

Reply:
Actually phobos unit tests doesn't give any deprecation because - apparently - test aren't done on range returning a struct as front (at least for "each" test).

So I can add a deprecation about this but both Ketmar and @Geod24 told me to be permissive instead. 

What to do?

Reply:
I guess it's best to open a thread in D.general about this. Imho if we can fix this properly without huge walls of deprecation messages, we should do so.
I also sent out an invitation to our D slack where we sometimes have discussions in #dmd too.

Reply:
On D.internals ketmar answer my question stone days ago. On slack geod24 said he agreed. I tried to open a discussion on general but I don't get so much attention. Anyway Steven Schveighoffer seemed to agree too.

See https://forum.dlang.org/post/pg0n5d$110$1@digitalmars.com

https://forum.dlang.org/post/b6f3b32debd34cc8883eec18827758d3@chiroptera